### PR TITLE
8325133: Missing MEMFLAGS parameter in parts of os API

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

Some parts of the `os` API calls NMT functions without taking a `MEMFLAGS` as a parameter. This PR extends the API of these functions to change this. No callers are changed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325133](https://bugs.openjdk.org/browse/JDK-8325133): Missing MEMFLAGS parameter in parts of os API (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17672/head:pull/17672` \
`$ git checkout pull/17672`

Update a local copy of the PR: \
`$ git checkout pull/17672` \
`$ git pull https://git.openjdk.org/jdk.git pull/17672/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17672`

View PR using the GUI difftool: \
`$ git pr show -t 17672`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17672.diff">https://git.openjdk.org/jdk/pull/17672.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17672#issuecomment-1921587708)

### CODE REVIEW

**Review by stefank — APPROVED** [review]
This seems fine to me. I might be good to think about pushing this further and remove some of the introduced default values.

File: src/hotspot/os/windows/os_windows.cpp [hunk 1]
```diff
@@ -3330,7 +3330,7 @@ char* os::replace_existing_mapping_with_file_mapping(char* base, size_t size, in
 // Multiple threads can race in this code but it's not possible to unmap small sections of
 // virtual space to get requested alignment, like posix-like os's.
 // Windows prevents multiple thread from remapping over each other so this loop is thread-safe.
-static char* map_or_reserve_memory_aligned(size_t size, size_t alignment, int file_desc) {
+static char* map_or_reserve_memory_aligned(size_t size, size_t alignment, int file_desc, MEMFLAGS flag = mtNone) {
```
Comment (gerard-ziemski) [inline src/hotspot/os/windows/os_windows.cpp:3333]: Wouldn't it be better NOT to implement a default value here?

If the point in this fix is to make sure that we are passing the NMT flags, then we should make sure that we review all the callers, and force them to define the flags explicitly.

File: src/hotspot/share/runtime/os.hpp [hunk 2]
```diff
@@ -442,7 +442,7 @@ class os: AllStatic {
 
   // Attempts to reserve the virtual memory at [addr, addr + bytes).
   // Does not overwrite existing mappings.
-  static char*  attempt_reserve_memory_at(char* addr, size_t bytes, bool executable = false);
+  static char*  attempt_reserve_memory_at(char* addr, size_t bytes, bool executable = false, MEMFLAGS flag = mtNone);
```
Comment (gerard-ziemski) [inline src/hotspot/share/runtime/os.hpp:445]: Same here, consider not implementing default value to force the callers to explicitly choose the value?

**Review by gerard-ziemski — APPROVED** [review]
Looks good.

I only have one feedback that frankly I myself am not 100% sure we need to address.

File: src/hotspot/share/runtime/os.hpp [hunk 3]
```diff
@@ -494,10 +494,10 @@ class os: AllStatic {
   static int create_file_for_heap(const char* dir);
   // Map memory to the file referred by fd. This function is slightly different from map_memory()
   // and is added to be used for implementation of -XX:AllocateHeapAt
-  static char* map_memory_to_file(size_t size, int fd);
-  static char* map_memory_to_file_aligned(size_t size, size_t alignment, int fd);
+  static char* map_memory_to_file(size_t size, int fd, MEMFLAGS flag = mtNone);
+  static char* map_memory_to_file_aligned(size_t size, size_t alignment, int fd, MEMFLAGS flag = mtNone);
   static char* map_memory_to_file(char* base, size_t size, int fd);
-  static char* attempt_map_memory_to_file_at(char* base, size_t size, int fd);
+  static char* attempt_map_memory_to_file_at(char* base, size_t size, int fd, MEMFLAGS flag = mtNone);
```
Comment (gerard-ziemski) [inline src/hotspot/share/runtime/os.hpp:500]: And here...

**Comment by jdksjolen** [comment]
Thank you for the reviews, I've answered your questions below.

/integrate

> Just curious if the PR is compiled and tested on all platforms?

No, but it works for all of our GHA tested platforms, which is sufficient for me. No way for me to check AIX, for example :-).

> Looks good.
> 
> I only have one feedback that frankly I myself am not 100% sure we need to address.



> This seems fine to me. I might be good to think about pushing this further and remove some of the introduced default values.

First off: Yes, we should not have default values for these functions. I am not going to fix for example `ReservedSpace` in this PR, that's for a future RFE. Second of all, introducing these default values is the same as what we do for `os::reserve_memory`, so this is established behavior, but not preferable behavior as mentioned.

**Pull request closed without merging**